x-pack/solutions/observability/plugins/apm_data_access/server/utils/unflatten_known_fields.ts

```diff
@@ -42,49 +42,39 @@ export function unflattenKnownApmEventFields(
   if (!hitFields) {
     return undefined;
   }
-  const missingRequiredFields =
-    requiredFields?.filter((key) => {
-      const value = hitFields?.[key];
-      return value === null || value === undefined || (Array.isArray(value) && value.length === 0);
-    }) ?? [];
 
-  if (missingRequiredFields.length > 0) {
-    throw new Error(`Missing required fields ${missingRequiredFields.join(', ')} in event`);
+  if (requiredFields) {
+    ensureRequiredApmFields(hitFields, requiredFields);
   }
 
-  const copy: Record<string, any> = mapToSingleOrMultiValue({
-    ...hitFields,
-  });
-
-  const [knownFields, unknownFields] = Object.entries(copy).reduce(
-    (prev, [key, value]) => {
-      if (ALL_FIELDS.has(key as KnownField)) {
-        prev[0][key as KnownField] = value;
-      } else {
-        prev[1][key] = value;
-      }
-      return prev;
-    },
-    [{} as Record<KnownField, any>, {} as Record<string, any>]
-  );
-
-  const unflattened = mergePlainObjects(
-    unflattenObject(unknownFields),
-    unflattenObject(knownFields)
-  );
+  // Merging still required in order to have correct overriding of known fields over unknown ones
+  const unflattened = mergePlainObjects(...unflattenSingleOrMultiValueFields(hitFields));
 
   return unflattened;
 }
 
-export function mapToSingleOrMultiValue<T extends Record<string, any>>(
-  fields: T
-): MapToSingleOrMultiValue<T> {
-  KNOWN_SINGLE_VALUED_FIELDS.forEach((field) => {
-    const value = fields[field];
-    if (value !== null && value !== undefined) {
-      fields[field as keyof T] = Array.isArray(value) ? value[0] : value;
-    }
-  });
+function unflattenSingleOrMultiValueFields<T extends Record<string, any>>(
+  fields: Readonly<T>
```

Comment:
We seem to be doing a lot of defensive copying, when we can just leverage the compiler to enforce no mutations of objects. We should make more use of `Readonly` in method signatures to keep us from mutating objects when we operate on them. That way, we can alleviate the need to always be copying/cloning objects, which when done a lot, creates a non-negligable overhead with memory usage and the GC.